lib/sycamore/sycamore/docset.py

```diff
@@ -987,6 +1011,157 @@ def llm_query(self, query_agent: LLMTextQueryAgent, **kwargs) -> "DocSet":
+            llm_cluster: Indicates whether an LLM should be used to normalize values of document field.
+            unique_field: Determines what makes a unique document.
+            **kwargs
+
+        Returns:
+            A DocSet with "properties.key" (unique values of document field)
+            and "properties.count" (frequency counts for unique values) which is
+            sorted based on descending and contains k records.
+        """
+
+        docset = self
+
+        if llm_cluster:
+            docset = docset.llm_cluster_entity(llm, description, field)
+            field = "properties._autogen_ClusterAssignment"
+
+        docset = docset.groupby_count(field, unique_field, **kwargs)
+
+        # uses 0 as default value -> end of docset
+        docset = docset.sort(descending, "properties.count", 0)
+        if k is not None:
+            docset = docset.limit(k)
+        return docset
+
+    def llm_cluster_entity(self, llm: LLM, description: str, field: str) -> "DocSet":
+        """
+        Normalizes a particular field of a DocSet. Identifies and assigns each document to a "group".
+
+        Args:
+            llm: LLM client.
+            description: Description of purpose of this operation.
+            field: Field to make/assign groups based on.
+
+        Returns:
+            A DocSet with an additional field "properties._autogen_ClusterAssignment" that contains
+            the assigned group.
+        """
+
+        docset = self
+        text = ""
+        for i, doc in enumerate(docset.take_all()):
+            if i != 0:
+                text += ", "
+            text += str(doc.field_to_value(field))
+
+        # sets message
+        messages = LlmClusterEntityFormGroupsMessagesPrompt(
+            field=field, description=description, text=text
+        ).get_messages_dict()
+
+        prompt_kwargs = {"messages": messages}
+
+        # call to LLM
+        completion = llm.generate(prompt_kwargs=prompt_kwargs, llm_kwargs={"temperature": 0})
+
+        groups = extract_json(completion)
+
+        assert isinstance(groups, dict)
+
+        # sets message
+        messagesForExtract = LlmClusterEntityAssignGroupsMessagesPrompt(
+            field=field, groups=groups["groups"]
+        ).get_messages_dict()
+
+        entity_extractor = OpenAIEntityExtractor(
+            entity_name="_autogen_ClusterAssignment",
+            llm=llm,
+            use_elements=False,
+            prompt=messagesForExtract,
+            field=field,
+        )
+        docset = docset.extract_entity(entity_extractor=entity_extractor)
+
+        # LLM response
+        return docset
+
+    def field_in(self, docset2: "DocSet", field1: str, field2: str) -> "DocSet":
+        """
+        Joins two docsets based on specified fields; docset1 filtered based on values of docset2.
+
+        SQL Equivalent: SELECT * FROM docset1 WHERE field1 IN (SELECT field2 FROM docset2);
```

Comment:
@bsowell and I wanted other people's opinion on the naming of this docset transform

Reply:
It's called `field_in`